src/box/memtx_space.c

```diff
@@ -1036,6 +1095,21 @@ memtx_build_on_replace(struct trigger *trigger, void *event)
 		if (stmt->old_tuple != NULL)
 			tuple_unref(stmt->old_tuple);
 	}
+	/*
+	 * Set on_rollback trigger on stmt to avoid
+	 * problem when rollbacked changes appears in
+	 * built-in-background index.
+	 */
+	struct on_rollback_trigger_with_data *on_rollback_associates =
+		region_aligned_alloc(&in_txn()->region,
+			sizeof(struct on_rollback_trigger_with_data),
+			alignof(struct on_rollback_trigger_with_data));
+	on_rollback_associates->data.stmt = stmt;
+	on_rollback_associates->data.state = state;
+	trigger_create(&on_rollback_associates->on_rollback,
+		       memtx_build_on_rollback, &on_rollback_associates->data,
+		       NULL);
+	txn_stmt_on_rollback(stmt, &on_rollback_associates->on_rollback);
```

Comment:
Creating such a trigger on each stmt may lead to significant memory consumption. Did you consider other options? For instance save in the one trigger up to several statements.

Reply:
The problem is rollback to savepoint: if we have a batch of stmt's, trigger will roll back the whole batch.

Reply:
Ok